### PR TITLE
Add python searcher library

### PR DESCRIPTION
**Pull request by buffalu**
Block Engine Client for Searchers in Python

### CODE REVIEW

File: python-searcher/README.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
+
+
+# About
+foo
```
Comment (0xEdgar) [inline python-searcher/README.md:4]: nice

File: python-searcher/python_searcher/__init__.py [hunk 2]
```diff
@@ -0,0 +1,2 @@
+from generated import *
+from searcher import JwtToken, SearcherInterceptor, get_searcher_client
```
Comment (0xEdgar) [inline python-searcher/python_searcher/__init__.py:2]: rename to `searcher_client` to keep consistent with rust path?

File: python-searcher/python_searcher/searcher-cli.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+    "--keypair-path",
+    help="Path to a keypair that is authenticated with the block engine.",
+    required=True,
+)
+@click.option(
+    "--block-engine-url",
+    help="Block Engine URL",
+    required=True,
+)
+def cli(
```
Comment (0xEdgar) [inline python-searcher/python_searcher/searcher-cli.py:33]: what formatter you using? i like https://github.com/psf/black

File: python-searcher/python_searcher/searcher-cli.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+    with open(payer) as kp_path:
+        payer_kp = Keypair.from_json(kp_path.read())
+
+    rpc_client = Client(rpc_url)
+    balance = rpc_client.get_balance(payer_kp.pubkey()).value
+    print(f"payer public key: {payer_kp.pubkey()} {balance=}")
+
+    is_leader_slot = False
+    print("waiting for jito leader...")
+    while not is_leader_slot:
```
Comment (0xEdgar) [inline python-searcher/python_searcher/searcher-cli.py:165]: if you wanna get fancy here u could use https://github.com/tqdm/tqdm for progress bars

File: python-searcher/python_searcher/searcher.py [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+        )
+
+    def _needs_authentication(self) -> bool:
+        """
+        The JWT needs authentication if there hasn't been an authentication yet or the access token has expired.
+        """
+        # return None in (
+        #     self._refresh_token,
+        #     self._access_token,
+        # ) or self._access_token.expiration > int(time.time())
```
Comment (0xEdgar) [inline python-searcher/python_searcher/searcher.py:138]: rm

File: python-searcher/python_searcher/searcher-cli.py (hunk 3 above)
Comment (buffalu) [reply python-searcher/python_searcher/searcher-cli.py:33]: black

File: python-searcher/python_searcher/searcher.py (hunk 5 above)
Comment (buffalu) [reply python-searcher/python_searcher/searcher.py:138]: still wip

File: python-searcher/python_searcher/searcher.py [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+        :param url: url of the Block Engine without http or https.
+        :param kp: block engine authentication keypair
+        """
+        self._url = url
+        self._kp = kp
+
+        self._access_token: Optional[JwtToken] = None
+        self._refresh_token: Optional[JwtToken] = None
+
+    def intercept_unary_stream(self, continuation, client_call_details, request):
```
[the remark this answers is not in the document]
Comment (buffalu) [reply python-searcher/python_searcher/searcher.py:66]: https://grpc.github.io/grpc/python/grpc.html#client-side-interceptor

File: python-searcher/README.md [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+# Tooling
+
+Install pip
+```asm
+$ curl -sSL https://bootstrap.pypa.io/get-pip.py | python 3 -
+```
+
+Install virtualenv
+```bash
+$ pip3 install virtualenv
```
Comment (esemeniuc) [inline python-searcher/README.md:15]: we can use venv, it is built in: https://docs.python.org/3/library/venv.html

File: python-searcher/python_searcher/searcher-cli.py [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+@click.option(
+    "--message",
+    help="Message in the bundle",
+    type=str,
+    required=True,
+)
+@click.option(
+    "--num_txs",
+    help="Number of transactions in the bundle (max is 5)",
+    type=str,
```
Comment (esemeniuc) [inline python-searcher/python_searcher/searcher-cli.py:129]: int?

File: python-searcher/python_searcher/searcher-cli.py [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+    required=True,
+)
+@click.option(
+    "--lamports",
+    help="Number of lamports to tip in each transaction",
+    type=str,
```
Comment (esemeniuc) [inline python-searcher/python_searcher/searcher-cli.py:135]: int?

File: python-searcher/python_searcher/searcher-cli.py [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+        time.sleep(0.5)
+        next_leader: NextScheduledLeaderResponse = client.GetNextScheduledLeader(
+            NextScheduledLeaderRequest()
+        )
+        num_slots_to_leader = next_leader.next_leader_slot - next_leader.current_slot
+        print(f"waiting {num_slots_to_leader} to jito leader")
+        is_leader_slot = num_slots_to_leader <= 2
+
+    blockhash = rpc_client.get_latest_blockhash().blockhash
+    txs = []
```
Comment (esemeniuc) [inline python-searcher/python_searcher/searcher-cli.py:175]: txs = [VersionedTransaction()]*num_txs

**Comment by esemeniuc** [comment]
Thinking about committing generated files
Pros:
- ease of setup (reduce time to first bundle)
- less platform specific breakage for installing codegen tooling
Cons:
- Git history bloat
- PR review noise

File: python-searcher/python_searcher/searcher-cli.py [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+from generated.searcher_pb2_grpc import SearcherServiceStub
+import click
+
+from solders.transaction import VersionedTransaction
+from solana.rpc.api import Client
+
+
+@click.group("cli")
+@click.pass_context
+@click.option(
```
Comment (esemeniuc) [inline python-searcher/python_searcher/searcher-cli.py:23]: can use Paths for better error handling: https://click.palletsprojects.com/en/8.1.x/arguments/#file-path-arguments

File: python-searcher/python_searcher/searcher.py [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+            client_call_details,
+        )
+
+        return continuation(client_call_details, request)
+
+    @staticmethod
+    def _insert_headers(
+        new_metadata: List[Tuple[str, str]], client_call_details
+    ) -> ClientCallDetails:
+        metadata = []
```
Comment (esemeniuc) [inline python-searcher/python_searcher/searcher.py:118]:
```suggestion
        metadata = (client_call_details.metadata or []) + new_metadata
```

File: python-searcher/python_searcher/searcher-cli.py (hunk 11 above)
Comment (esemeniuc) [reply python-searcher/python_searcher/searcher-cli.py:23]: There's also an explicit File type if that works better: https://click.palletsprojects.com/en/8.1.x/api/#click.File

**Comment by esemeniuc** [comment]
I'd rename python-searcher to `python` like rust, but not super important

File: README.md [hunk 13]
```diff
@@ -3,12 +3,13 @@ This repository contains a few examples on how to use a searcher.
...
 ## Setup
 - You may have to install protoc to compile protobufs.
+
 - Ensure that rust is installed:
 ```bash
 $ curl https://sh.rustup.rs -sSf | sh
 $ source $HOME/.cargo/env
 $ git clone https://github.com/jito-labs/searcher-examples.git
-$ cd searcher-examples
+$ cd jito_searcher_client-examples
```
Comment (segfaultdoc) [inline README.md:12]: this wrong

File: python/jito_searcher_client/generated/relayer_pb2_grpc.pyi [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+    """
+    SubscribePackets: grpc.UnaryStreamMultiCallable[
+        relayer_pb2.SubscribePacketsRequest,
+        relayer_pb2.SubscribePacketsResponse,
+    ]
+    """Validators can subscribe to packets from the relayer and receive a multiplexed signal that contains a mixture
+    of packets and heartbeats
+    """
+
+class RelayerServicer(metaclass=abc.ABCMeta):
```
Comment (segfaultdoc) [inline python/jito_searcher_client/generated/relayer_pb2_grpc.pyi:33]: why do we need relayer code in this repo?

**Review by segfaultdoc — APPROVED** [review]

**Comment by buffalu** [comment]
going to move to python repo

**Pull request closed without merging**